File: 3445-lexicographically-minimum-string-after-removing-stars/lexicographically-minimum-string-after-removing-stars.go
// Priority Queue để lưu ký tự theo thứ tự từ điển (min-heap)
type CharHeap []byte;

func (h CharHeap) Len() int            { return len(h); }
func (h CharHeap) Less(i, j int) bool  { return h[i] < h[j]; }
func (h CharHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i]; }
func (h *CharHeap) Push(x interface{}) { *h = append(*h, x.(byte)); }
func (h *CharHeap) Pop() interface{} {
	old := *h;
	n := len(old);
	x := old[n-1];
	*h = old[:n-1];
	return x;
}

func clearStars(s string) string {
	n := len(s);
	keep := make([]bool, n); // keep[i] = true nếu giữ lại ký tự s[i]
	for i := 0; i < n; i++ {
		keep[i] = true;
	}

	h := &CharHeap{};
	heap.Init(h);

	// map ký tự -> danh sách các index (theo thứ tự xuất hiện)
	positions := make(map[byte][]int);

	for i := 0; i < n; i++ {
		ch := s[i];
		if ch == '*' {
			// Lấy ký tự nhỏ nhất
			if h.Len() == 0 {
				continue; // nếu không có ký tự nào còn lại để xóa, bỏ qua
			}
			smallest := heap.Pop(h).(byte);

			// Lấy index cuối cùng của ký tự nhỏ nhất (gần nhất với dấu *)
			idxList := positions[smallest];
			indexToRemove := idxList[len(idxList)-1];
			positions[smallest] = idxList[:len(idxList)-1];

			// Đánh dấu không giữ lại
			keep[i] = false;               // bỏ dấu '*'
			keep[indexToRemove] = false;  // bỏ ký tự nhỏ nhất trước đó
		} else {
			heap.Push(h, ch);
			positions[ch] = append(positions[ch], i);
		}
	}

	// Duyệt và xây kết quả
	result := []byte{};
	for i := 0; i < n; i++ {
		if keep[i] {
			result = append(result, s[i]);
		}
	}
	return string(result);
}

File: 3445-lexicographically-minimum-string-after-removing-stars/lexicographically-minimum-string-after-removing-stars.java
class Solution {
    public String clearStars(String s) {
        int n = s.length();

        // Dùng mảng boolean để đánh dấu ký tự nào sẽ được giữ lại
        boolean[] keep = new boolean[n];
        Arrays.fill(keep, true);

        // PriorityQueue lưu trữ các ký tự đã xuất hiện, để tìm ký tự nhỏ nhất (theo thứ tự từ điển)
        PriorityQueue<Character> pq = new PriorityQueue<>();

        // Map ký tự → danh sách các vị trí xuất hiện của nó (để xóa đúng ký tự nhỏ nhất)
        Map<Character, Deque<Integer>> charIndices = new HashMap<>();

        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);

            if (c == '*') {
                // Lấy ra ký tự nhỏ nhất hiện tại (nằm bên trái '*')
                char smallestChar = pq.poll();

                // Lấy index gần nhất của ký tự nhỏ nhất (từ bên phải sang trái)
                int indexToRemove = charIndices.get(smallestChar).removeLast();

                // Đánh dấu bỏ cả dấu '*' và ký tự nhỏ nhất
                keep[i] = false; // bỏ '*'
                keep[indexToRemove] = false; // bỏ ký tự nhỏ nhất
            } else {
                // Đưa ký tự vào PriorityQueue và lưu vị trí nó xuất hiện
                pq.offer(c);
                charIndices.putIfAbsent(c, new ArrayDeque<>());
                charIndices.get(c).add(i);
            }
        }

        // Xây dựng chuỗi kết quả chỉ với các ký tự được giữ lại
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < n; i++)
            if (keep[i])
                result.append(s.charAt(i));

        return result.toString();
    }
}
